AI autofix editor summary

Changes made:
- `scripts/review_rb_judge.sh`: replaced the broken pipe+heredoc truncation command with a `python3 -c` form so the piped `PR_DIFF` bytes actually reach the UTF-8-safe truncation logic and ShellCheck SC2259 is eliminated.
- `scripts/review_rb_judge.sh`: moved the prompt-size diagnostic to run after the first `sanitize_codex_prompt_file "${RB_JUDGE_PROMPT}"` call, so the logged size and warning reflect the sanitized prompt Codex actually reads.
- Verification: `bash -n scripts/review_rb_judge.sh` passed; `shellcheck --severity=error scripts/review_rb_judge.sh` passed; `git diff --check -- scripts/review_rb_judge.sh` passed; `git diff --stat -- scripts/review_rb_judge.sh` reports `1 file changed, 16 insertions(+), 13 deletions(-)`.
- Assumption: keeping the existing `RB_JUDGE_PR_DIFF_MAX_BYTES=400000` default and `950000` warning threshold unchanged is the lowest-risk scoped fix because the verified defects were the stdin-redirection bug and the pre-sanitization diagnostic.
- Missing context: the CI snapshot exposed the lint failure directly, but there was no separate large-PR runtime reproduction in this checkout to exercise the judge path end-to-end after the fix.

Change status:
- edited

Already satisfied (suggested but already present):
- Positive-integer validation/fallback for `RB_JUDGE_PR_DIFF_MAX_BYTES` was already present in `scripts/review_rb_judge.sh`.
- Byte-accurate diff sizing/truncation was already present in `scripts/review_rb_judge.sh` via `wc -c` plus the UTF-8-safe byte-prefix logic, so the older bot comments about `${#PR_DIFF}` no longer applied.

Ignored suggestions (with short reason):
- Copilot review entry 1 low-confidence bytes-vs-characters warning — ignored because it was advisory-only, not tied to a concrete failure, and the post-sanitization byte measurement is still a conservative near-cap diagnostic.
- Copilot review entry 3 (`scripts/review_rb_judge.sh:338`) — ignored because switching this path to `_embed_input_file`/diff-boundary truncation would be a broader refactor, and no reviewer or CI evidence showed the current byte-prefix truncation breaks judge execution.
- FLOOR tag `scripts/review_rb_judge.sh:322` — ignored because the code already coerces an empty `wc -c` result to `0` before the numeric comparison, so there is no remaining arithmetic failure path.

Reviewer files processed:
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_deepseek_deepseek-v4-pro.txt` | checksum: `056368721bd8fb65d36415d6891facdd8212b9ea38712901aa5089122ca3bfb6` | used — supplied the validated SC2259 bug and the prompt-size diagnostic gap.
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_minimax_minimax-m2_5.txt` | checksum: `c45123ec8c78273417c3281a03a64c908583cbee8625d17801e718b122c30466` | used — independently confirmed the blocking SC2259 lint failure.
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_mistralai_mistral-small-2603.txt` | checksum: `e456b7943785b184dc7ecb79e99198aa029759ab697d1b3514e0b354e4d253c0` | ignored — contained no actionable findings.
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_moonshotai_kimi-k2_5.txt` | checksum: `5f667d8b4f645b813654c247c1dc0fa3697497753c448431ce044fca92134d56` | used — confirmed both the stdin conflict and the diagnostic-order issue.
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_qwen_qwen3_6-plus.txt` | checksum: `ee34e06117bbc4126d50f786918be3636b78677632e3613d4dbf4b5e5ee78e83` | used — provided detailed validation for the same two real issues.
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_x-ai_grok-4_20.txt` | checksum: `a8333e4f319e5abdaa54fa32042e2f323010150a7618f75bc808d3cfd16043c2` | used — confirmed the runtime impact of the SC2259 regression and the observability gap.

Review file issue audit:
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_deepseek_deepseek-v4-pro.txt` | total issues listed: 2 | issues applied: 2 | issues already applied: 0 | issues ignored: 0
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_minimax_minimax-m2_5.txt` | total issues listed: 1 | issues applied: 1 | issues already applied: 0 | issues ignored: 0
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_mistralai_mistral-small-2603.txt` | total issues listed: 0 | issues applied: 0 | issues already applied: 0 | issues ignored: 0
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_moonshotai_kimi-k2_5.txt` | total issues listed: 2 | issues applied: 2 | issues already applied: 0 | issues ignored: 0
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_qwen_qwen3_6-plus.txt` | total issues listed: 3 | issues applied: 3 | issues already applied: 0 | issues ignored: 0
- `/tmp/codex-pr-26096166544-1-13734/previous_reviews/review_x-ai_grok-4_20.txt` | total issues listed: 2 | issues applied: 2 | issues already applied: 0 | issues ignored: 0

PR comment audit:
- entry[0] `Copilot` | `scripts/review_rb_judge.sh:319` | disposition: already satisfied | the code already uses byte-accurate sizing/truncation instead of `${#PR_DIFF}`.
- entry[1] `copilot-pull-request-reviewer[bot]` | path/line: n/a | disposition: ignored | the suppressed units warning was low-confidence and not tied to a concrete failure after moving the diagnostic post-sanitization.
- entry[2] `Copilot` | `scripts/review_rb_judge.sh:319` | disposition: already satisfied | positive-integer validation and fallback for `RB_JUDGE_PR_DIFF_MAX_BYTES` were already present.
- entry[3] `Copilot` | `scripts/review_rb_judge.sh:338` | disposition: ignored | diff-boundary truncation would require a broader refactor than the verified fix, and no concrete failure showed the current truncation note/path is incorrect.
- entry[4] `gemini-code-assist[bot]` | path/line: n/a | disposition: already satisfied | the review summary's concerns were resolved by existing byte-counting plus the already-present cap validation.
- entry[5] `gemini-code-assist[bot]` | `scripts/review_rb_judge.sh:318` | disposition: already satisfied | the code no longer uses `${#PR_DIFF}` and already validates the cap before arithmetic comparison.

Regression fingerprint:
- `scripts/review_rb_judge.sh:324-336` — prior autofix introduced a `printf ... | python3 - ... <<'PY'` pipe+heredoc combination, producing ShellCheck SC2259 and preventing the truncation path from consuming `PR_DIFF`.
- `scripts/review_rb_judge.sh:555-583` — prior autofix logged prompt size before `sanitize_codex_prompt_file`, so the warning could describe a different payload than the one actually sent to Codex.

Runtime failure path:
- `lint` check-run -> `shellcheck --severity=error scripts/review_rb_judge.sh` -> SC2259 at the diff-truncation command -> workflow fails before `review_rb_judge.sh` can run.
- `review_rb_judge.sh` attempt loop -> `sanitize_codex_prompt_file "${RB_JUDGE_PROMPT}"` mutates the prompt after the old diagnostic point -> logged size/warning can diverge from the actual stdin payload used by `codex ... < "${RB_JUDGE_PROMPT}"`.

Files changed / commit status:
- scripts/review_rb_judge.sh

--- a/scripts/review_rb_judge.sh
+++ b/scripts/review_rb_judge.sh
@@ -38,13 +38,75 @@ if ! command -v sanitize_codex_prompt_file >/dev/null 2>&1; then
   sanitize_codex_prompt_file() { :; }
 fi
 if ! command -v _embed_input_file >/dev/null 2>&1; then
-  _init_prompt_budget() { :; }
-  _cleanup_prompt_budget() { :; }
+  : "${_PROMPT_BUDGET_TOTAL_BYTES:=800000}"
+  _prompt_budget_state_file() {
+    printf '%s/_prompt_input_budget_state.%s\n' "${TMPDIR:-/tmp}" "$$"
+  }
+  _init_prompt_budget() {
+    local _cap="${1:-${_PROMPT_BUDGET_TOTAL_BYTES}}"
+    local _state
+    _state="$(_prompt_budget_state_file)"
+    export _PROMPT_BUDGET_TOTAL_BYTES="${_cap}"
+    printf '0\n' > "${_state}" 2>/dev/null || true
+  }
+  _cleanup_prompt_budget() {
+    local _state
+    _state="$(_prompt_budget_state_file)"
+    rm -f "${_state}" 2>/dev/null || true
+  }
   _embed_input_file() {
     local _p="${1:-}"
+    local _cap="${2:-100000}"
+    local _mode="${3:-head}"
+    local _state _used _budget_remaining _effective_cap _size _emit_bytes
     if [ -z "${_p}" ] || [ ! -e "${_p}" ]; then printf '(missing)\n'; return 0; fi
     if [ ! -s "${_p}" ]; then printf '(empty)\n'; return 0; fi
-    cat "${_p}"
+
+    _state="$(_prompt_budget_state_file)"
+    _used=0
+    if [ -f "${_state}" ]; then
+      _used="$(cat "${_state}" 2>/dev/null)"
+      [ -n "${_used}" ] || _used=0
+    fi
+    _budget_remaining=$(( _PROMPT_BUDGET_TOTAL_BYTES - _used ))
+    if [ "${_budget_remaining}" -le 0 ]; then
+      printf '(omitted — total prompt input budget %d bytes exhausted; %d bytes already inlined)\n' \
+        "${_PROMPT_BUDGET_TOTAL_BYTES}" "${_used}"
+      return 0
+    fi
+
+    _effective_cap="${_cap}"
+    if [ "${_budget_remaining}" -lt "${_effective_cap}" ]; then
+      _effective_cap="${_budget_remaining}"
+    fi
+
+    _size="$(wc -c < "${_p}" 2>/dev/null | tr -d '[:space:]')"
+    [ -n "${_size}" ] || _size=0
+    if [ "${_size}" -le "${_effective_cap}" ]; then
+      cat "${_p}"
+      _emit_bytes="${_size}"
+    else
+      PYTHONDONTWRITEBYTECODE=1 python3 - "${_p}" "${_effective_cap}" 2>/dev/null <<'PY' || head -c "${_effective_cap}" "${_p}"
+from pathlib import Path
+import sys
+
+data = Path(sys.argv[1]).read_bytes()
+cap = int(sys.argv[2])
+if cap > 0 and len(data) > cap:
+    i = cap
+    while i > 0 and (data[i] & 0xC0) == 0x80:
+        i -= 1
+    data = data[:i]
+sys.stdout.buffer.write(data)
+PY
+      printf '\n[... TRUNCATED — file is %s bytes; first %s bytes shown above (mode=%s, per-file cap=%s, budget remaining was %s; fallback embedder) ...]\n' \
+        "${_size}" "${_effective_cap}" "${_mode}" "${_cap}" "${_budget_remaining}"
+      _emit_bytes="${_effective_cap}"
+    fi
+
+    if [ -n "${_emit_bytes}" ] && [ "${_emit_bytes}" -gt 0 ] 2>/dev/null; then
+      printf '%s\n' "$(( _used + _emit_bytes ))" > "${_state}" 2>/dev/null || true
+    fi
   }
 fi
 REVIEW_RB_SEMBLE_HELPERS_AVAILABLE="false"
@@ -322,7 +384,8 @@ PR_DIFF="$(gh_retry gh api "repos/${REPOSITORY}/pulls/${PR_NUMBER}" \
 # `head -1000` truncation removed in PR #2564 (commit ebe2ecf3) was
 # the last guard here; this restores one at a byte budget instead
 # of a line budget so the cap is predictable across diff styles.
-# Override via the RB_JUDGE_PR_DIFF_MAX_BYTES repo var.
+# Override via the RB_JUDGE_PR_DIFF_MAX_BYTES environment variable
+# (for example, export a repository variable into env in the caller).
 RB_JUDGE_PR_DIFF_MAX_BYTES="${RB_JUDGE_PR_DIFF_MAX_BYTES:-400000}"
 if ! [[ "${RB_JUDGE_PR_DIFF_MAX_BYTES}" =~ ^[0-9]+$ ]] || [ "${RB_JUDGE_PR_DIFF_MAX_BYTES}" -le 0 ]; then
   echo "::warning::Invalid RB_JUDGE_PR_DIFF_MAX_BYTES='${RB_JUDGE_PR_DIFF_MAX_BYTES}'; using 400000."
@@ -381,7 +444,7 @@ RB_JUDGE_PR_META_RENDER_FILE="${RUNTIME_DIR}/rb_judge_pr_meta.json"
 RB_JUDGE_PR_COMMENTS_RENDER_FILE="${RUNTIME_DIR}/rb_judge_pr_comments.json"
 RB_JUDGE_PR_REVIEW_COMMENTS_RENDER_FILE="${RUNTIME_DIR}/rb_judge_pr_review_comments.json"
 RB_JUDGE_SEMBLE_PREFETCH=""
-trap 'rm -f "${RB_JUDGE_SEMBLE_QUERY_FILE:-}" "${RB_JUDGE_REQUIREMENT_FILE:-}" "${RB_JUDGE_PR_META_RENDER_FILE:-}" "${RB_JUDGE_PR_COMMENTS_RENDER_FILE:-}" "${RB_JUDGE_PR_REVIEW_COMMENTS_RENDER_FILE:-}"' EXIT
+trap '_cleanup_prompt_budget; rm -f "${RB_JUDGE_SEMBLE_QUERY_FILE:-}" "${RB_JUDGE_REQUIREMENT_FILE:-}" "${RB_JUDGE_PR_META_RENDER_FILE:-}" "${RB_JUDGE_PR_COMMENTS_RENDER_FILE:-}" "${RB_JUDGE_PR_REVIEW_COMMENTS_RENDER_FILE:-}"' EXIT
 
 {
   printf '%s\n' 'Review-blocked judge context.'
@@ -404,10 +467,14 @@ RB_JUDGE_SEMBLE_PREFETCH="$(render_review_rb_semble_prefetch "${RB_JUDGE_SEMBLE_
 # Keep the non-diff judge inputs under a shared budget so oversized PR
 # discussions / inline reviews cannot reintroduce the same `turn/start`
 # prompt-envelope failure the PR diff cap was added to prevent.
-if [ -n "${FIRST_ISSUE}" ]; then
+if [ -n "${FIRST_ISSUE}" ] && [ -n "${FIRST_ISSUE_BODY//[[:space:]]/}" ]; then
   printf '%s\n' "${FIRST_ISSUE_BODY}" > "${RB_JUDGE_REQUIREMENT_FILE}"
 else
   {
+    if [ -n "${FIRST_ISSUE}" ]; then
+      echo "[NOTE: linked issue #${FIRST_ISSUE} has no body; using PR title/body as requirement fallback.]"
+      echo
+    fi
     echo "Title: $(jq -r '.title // ""' "${PR_META_FILE}")"
     echo "Body: $(jq -r '.body // ""' "${PR_PAYLOAD_FILE}")"
   } > "${RB_JUDGE_REQUIREMENT_FILE}"
@@ -478,13 +545,15 @@ _init_prompt_budget "${RB_JUDGE_CONTEXT_BUDGET_BYTES}"
   fi
   printf '%s\n' "${PR_DIFF}"
   echo
-  echo "=== PR #${PR_NUMBER} COMMENTS (editor summaries, reviewer findings) ==="
-  echo
-  _embed_input_file "${RB_JUDGE_PR_COMMENTS_RENDER_FILE}" "${RB_JUDGE_PR_COMMENTS_MAX_BYTES}"
-  echo
   echo "=== PR #${PR_NUMBER} INLINE REVIEW COMMENTS ==="
   echo
   _embed_input_file "${RB_JUDGE_PR_REVIEW_COMMENTS_RENDER_FILE}" "${RB_JUDGE_PR_REVIEW_COMMENTS_MAX_BYTES}"
+  echo
+  # Keep the file/line reviewer findings ahead of general PR discussion
+  # so the shared prompt budget preserves the most actionable evidence.
+  echo "=== PR #${PR_NUMBER} COMMENTS (editor summaries, reviewer findings) ==="
+  echo
+  _embed_input_file "${RB_JUDGE_PR_COMMENTS_RENDER_FILE}" "${RB_JUDGE_PR_COMMENTS_MAX_BYTES}"
   echo
   echo "=== REVIEW-BLOCKED CONTEXT ==="
   echo "Review-blocked judge retry: $((RETRY_COUNT + 1)) of ${MAX_REVIEW_BLOCKED_RETRIES}"

--- a/scripts/review_rb_judge.sh
+++ b/scripts/review_rb_judge.sh
@@ -395,7 +395,8 @@ PR_DIFF_TRUNCATED=false
 PR_DIFF_BYTES_TOTAL="$(printf '%s' "${PR_DIFF}" | wc -c | tr -cd '0-9' || true)"
 [ -n "${PR_DIFF_BYTES_TOTAL}" ] || PR_DIFF_BYTES_TOTAL=0
 if [ "${PR_DIFF_BYTES_TOTAL}" -gt "${RB_JUDGE_PR_DIFF_MAX_BYTES}" ]; then
-  PR_DIFF="$(printf '%s' "${PR_DIFF}" | PYTHONDONTWRITEBYTECODE=1 python3 -c '
+  PR_DIFF_TRUNCATED_CONTENT=""
+  if PR_DIFF_TRUNCATED_CONTENT="$(printf '%s' "${PR_DIFF}" | PYTHONDONTWRITEBYTECODE=1 python3 -c '
 import sys
 
 data = sys.stdin.buffer.read()
@@ -406,7 +407,17 @@ if cap > 0 and len(data) > cap:
         i -= 1
     data = data[:i]
 sys.stdout.buffer.write(data)
-' "${RB_JUDGE_PR_DIFF_MAX_BYTES}")"
+' "${RB_JUDGE_PR_DIFF_MAX_BYTES}" 2>/dev/null)"; then
+    PR_DIFF="${PR_DIFF_TRUNCATED_CONTENT}"
+  else
+    echo "::warning::python3 unavailable for UTF-8-safe PR diff truncation; falling back to a raw byte prefix. sanitize_codex_prompt_file will strip any invalid trailing bytes before codex reads the prompt."
+    PR_DIFF_FALLBACK_FILE="$(mktemp "${TMPDIR:-/tmp}/review-rb-judge-pr-diff.XXXXXX")"
+    printf '%s' "${PR_DIFF}" > "${PR_DIFF_FALLBACK_FILE}"
+    PR_DIFF="$(head -c "${RB_JUDGE_PR_DIFF_MAX_BYTES}" "${PR_DIFF_FALLBACK_FILE}")"
+    rm -f "${PR_DIFF_FALLBACK_FILE}" 2>/dev/null || true
+    unset PR_DIFF_FALLBACK_FILE
+  fi
+  unset PR_DIFF_TRUNCATED_CONTENT
   PR_DIFF_TRUNCATED=true
 fi
 PRELOADED_PR_META="$(jq -c '{
@@ -540,7 +551,7 @@ _init_prompt_budget "${RB_JUDGE_CONTEXT_BUDGET_BYTES}"
   # the model runs and every retry in the reasoning ladder fails
   # identically (see PR shubhodeep1/bitsafe.io#368, run 26092826715).
   if [ "${PR_DIFF_TRUNCATED}" = "true" ]; then
-    echo "[NOTE: PR diff is ${PR_DIFF_BYTES_TOTAL} bytes; truncated to a UTF-8-safe prefix within ${RB_JUDGE_PR_DIFF_MAX_BYTES} bytes to fit codex stdin (1 MB cap). Use exec-read on specific files for the elided tail if needed; the judge runs --sandbox read-only so file reads are available.]"
+    echo "[NOTE: PR diff is ${PR_DIFF_BYTES_TOTAL} bytes; truncated to a prefix within ${RB_JUDGE_PR_DIFF_MAX_BYTES} bytes to fit codex stdin (1 MB cap). Use exec-read on specific files for the elided tail if needed; the judge runs --sandbox read-only so file reads are available.]"
     echo
   fi
   printf '%s\n' "${PR_DIFF}"

--- a/scripts/review_rb_judge.sh
+++ b/scripts/review_rb_judge.sh
@@ -37,6 +37,16 @@ fi
 if ! command -v sanitize_codex_prompt_file >/dev/null 2>&1; then
   sanitize_codex_prompt_file() { :; }
 fi
+if ! command -v _embed_input_file >/dev/null 2>&1; then
+  _init_prompt_budget() { :; }
+  _cleanup_prompt_budget() { :; }
+  _embed_input_file() {
+    local _p="${1:-}"
+    if [ -z "${_p}" ] || [ ! -e "${_p}" ]; then printf '(missing)\n'; return 0; fi
+    if [ ! -s "${_p}" ]; then printf '(empty)\n'; return 0; fi
+    cat "${_p}"
+  }
+fi
 REVIEW_RB_SEMBLE_HELPERS_AVAILABLE="false"
 REVIEW_RB_SEMBLE_MAX_CHUNKS="4"
 REVIEW_RB_SEMBLE_QUERY_MAX_BYTES="12000"
@@ -366,8 +376,12 @@ fi
 RB_JUDGE_PROMPT="${RUNTIME_DIR}/rb_judge_prompt.txt"
 RB_JUDGE_OUTPUT="${RUNTIME_DIR}/rb_judge_output.txt"
 RB_JUDGE_SEMBLE_QUERY_FILE="${RUNTIME_DIR}/rb_judge_semble_query.txt"
+RB_JUDGE_REQUIREMENT_FILE="${RUNTIME_DIR}/rb_judge_requirement.txt"
+RB_JUDGE_PR_META_RENDER_FILE="${RUNTIME_DIR}/rb_judge_pr_meta.json"
+RB_JUDGE_PR_COMMENTS_RENDER_FILE="${RUNTIME_DIR}/rb_judge_pr_comments.json"
+RB_JUDGE_PR_REVIEW_COMMENTS_RENDER_FILE="${RUNTIME_DIR}/rb_judge_pr_review_comments.json"
 RB_JUDGE_SEMBLE_PREFETCH=""
-trap 'rm -f "${RB_JUDGE_SEMBLE_QUERY_FILE:-}"' EXIT
+trap 'rm -f "${RB_JUDGE_SEMBLE_QUERY_FILE:-}" "${RB_JUDGE_REQUIREMENT_FILE:-}" "${RB_JUDGE_PR_META_RENDER_FILE:-}" "${RB_JUDGE_PR_COMMENTS_RENDER_FILE:-}" "${RB_JUDGE_PR_REVIEW_COMMENTS_RENDER_FILE:-}"' EXIT
 
 {
   printf '%s\n' 'Review-blocked judge context.'
@@ -386,6 +400,33 @@ Body: $(jq -r '.body // ""' "${PR_PAYLOAD_FILE}")" \
   append_review_rb_semble_query_section "PR review comments JSON:" "${PR_REVIEW_COMMENTS}" 2500
 } > "${RB_JUDGE_SEMBLE_QUERY_FILE}"
 RB_JUDGE_SEMBLE_PREFETCH="$(render_review_rb_semble_prefetch "${RB_JUDGE_SEMBLE_QUERY_FILE}" "Review-Blocked Judge Context")"
+
+# Keep the non-diff judge inputs under a shared budget so oversized PR
+# discussions / inline reviews cannot reintroduce the same `turn/start`
+# prompt-envelope failure the PR diff cap was added to prevent.
+if [ -n "${FIRST_ISSUE}" ]; then
+  printf '%s\n' "${FIRST_ISSUE_BODY}" > "${RB_JUDGE_REQUIREMENT_FILE}"
+else
+  {
+    echo "Title: $(jq -r '.title // ""' "${PR_META_FILE}")"
+    echo "Body: $(jq -r '.body // ""' "${PR_PAYLOAD_FILE}")"
+  } > "${RB_JUDGE_REQUIREMENT_FILE}"
+fi
+if ! printf '%s\n' "${PR_META_JSON}" | jq '.' > "${RB_JUDGE_PR_META_RENDER_FILE}" 2>/dev/null; then
+  printf '%s\n' "${PR_META_JSON}" > "${RB_JUDGE_PR_META_RENDER_FILE}"
+fi
+if ! printf '%s\n' "${PR_COMMENTS}" | jq '.' > "${RB_JUDGE_PR_COMMENTS_RENDER_FILE}" 2>/dev/null; then
+  printf '%s\n' "${PR_COMMENTS}" > "${RB_JUDGE_PR_COMMENTS_RENDER_FILE}"
+fi
+if ! printf '%s\n' "${PR_REVIEW_COMMENTS}" | jq '.' > "${RB_JUDGE_PR_REVIEW_COMMENTS_RENDER_FILE}" 2>/dev/null; then
+  printf '%s\n' "${PR_REVIEW_COMMENTS}" > "${RB_JUDGE_PR_REVIEW_COMMENTS_RENDER_FILE}"
+fi
+RB_JUDGE_CONTEXT_BUDGET_BYTES=300000
+RB_JUDGE_REQUIREMENT_MAX_BYTES=50000
+RB_JUDGE_PR_META_MAX_BYTES=50000
+RB_JUDGE_PR_COMMENTS_MAX_BYTES=150000
+RB_JUDGE_PR_REVIEW_COMMENTS_MAX_BYTES=100000
+_init_prompt_budget "${RB_JUDGE_CONTEXT_BUDGET_BYTES}"
 
 {
   if [ -f ./pre_assembled_static.txt ]; then
@@ -408,17 +449,16 @@ RB_JUDGE_SEMBLE_PREFETCH="$(render_review_rb_semble_prefetch "${RB_JUDGE_SEMBLE_
   if [ -n "${FIRST_ISSUE}" ]; then
     echo "=== ISSUE #${FIRST_ISSUE} (original requirement) ==="
     echo
-    echo "${FIRST_ISSUE_BODY}"
+    _embed_input_file "${RB_JUDGE_REQUIREMENT_FILE}" "${RB_JUDGE_REQUIREMENT_MAX_BYTES}"
   else
     echo "=== PR DESCRIPTION (no linked issue) ==="
     echo
-    echo "Title: $(jq -r '.title // ""' "${PR_META_FILE}")"
-    echo "Body: $(jq -r '.body // ""' "${PR_PAYLOAD_FILE}")"
+    _embed_input_file "${RB_JUDGE_REQUIREMENT_FILE}" "${RB_JUDGE_REQUIREMENT_MAX_BYTES}"
   fi
   echo
   echo "=== PR #${PR_NUMBER} METADATA ==="
   echo
-  printf '%s\n' "${PR_META_JSON}" | jq '.'
+  _embed_input_file "${RB_JUDGE_PR_META_RENDER_FILE}" "${RB_JUDGE_PR_META_MAX_BYTES}"
   echo
   echo "=== PR #${PR_NUMBER} DIFF ==="
   echo
@@ -440,11 +480,11 @@ RB_JUDGE_SEMBLE_PREFETCH="$(render_review_rb_semble_prefetch "${RB_JUDGE_SEMBLE_
   echo
   echo "=== PR #${PR_NUMBER} COMMENTS (editor summaries, reviewer findings) ==="
   echo
-  printf '%s\n' "${PR_COMMENTS}" | jq '.'
+  _embed_input_file "${RB_JUDGE_PR_COMMENTS_RENDER_FILE}" "${RB_JUDGE_PR_COMMENTS_MAX_BYTES}"
   echo
   echo "=== PR #${PR_NUMBER} INLINE REVIEW COMMENTS ==="
   echo
-  printf '%s\n' "${PR_REVIEW_COMMENTS}" | jq '.'
+  _embed_input_file "${RB_JUDGE_PR_REVIEW_COMMENTS_RENDER_FILE}" "${RB_JUDGE_PR_REVIEW_COMMENTS_MAX_BYTES}"
   echo
   echo "=== REVIEW-BLOCKED CONTEXT ==="
   echo "Review-blocked judge retry: $((RETRY_COUNT + 1)) of ${MAX_REVIEW_BLOCKED_RETRIES}"
@@ -471,6 +511,7 @@ RB_JUDGE_SEMBLE_PREFETCH="$(render_review_rb_semble_prefetch "${RB_JUDGE_SEMBLE_
     echo "the PR's work should be discarded."
   fi
 } > "${RB_JUDGE_PROMPT}"
+_cleanup_prompt_budget
 rm -f "${RB_JUDGE_SEMBLE_QUERY_FILE}"
 
 # -----------------------------------------------------------

